### PR TITLE
Acquisition

### PR DESCRIPTION
**Pull request by fielia**

### CODE REVIEW

File: src/main/java/frc/robot/RobotMap.java [hunk 1]
```diff
@@ -13,4 +13,13 @@ public class Control {
 		public static final int LEGACY_GAMEPAD = 0;
 	}
 
+	/**
+	 * Acquisition Motor ports
+	 */
+	public class Acquisition {
+		public static final int LIMIT_SWITCH = 2;
```
Comment (vladaviedov) [inline src/main/java/frc/robot/RobotMap.java:20]: Can you add a comment that this is a PWM port and not a CAN port?

File: src/main/java/frc/robot/commands/manual/ManualAcquisition.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import frc.robot.subsystems.Acquisition;
+
+public class ManualAcquisition extends CommandBase {
+	private Acquisition _acquisition;
+	private IControlInput _controls;
+	// FIXME: set fields to the proper value
+	private static final int SPIN_SPEED = 0;
+	private static final int TILT_SPEED = 0;
+
+	public ManualAcquisition(Acquisition _acq, IControlInput _controlInput) {
```
Comment (vladaviedov) [inline src/main/java/frc/robot/commands/manual/ManualAcquisition.java:14]: Underscores are reserved for class fields, don't put them in function args

File: src/main/java/frc/robot/subsystems/Acquisition.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import frc.robot.RobotMap;
+
+public class Acquisition extends SubsystemBase {
+	// motor and limit switch variables
+	private CANSparkMax _spinMotor = new CANSparkMax(RobotMap.Acquisition.SPIN_MOTOR_PORT, MotorType.kBrushless);
+	private CANSparkMax _tiltMotor = new CANSparkMax(RobotMap.Acquisition.TILT_MOTOR_PORT, MotorType.kBrushed);
+	private CANSparkMax _ballChannelMotor = new CANSparkMax(RobotMap.Acquisition.BALL_CHANNEL_MOTOR_PORT, MotorType.kBrushless);
+	private DigitalInput _tiltLimitSwitch = new DigitalInput(RobotMap.Acquisition.LIMIT_SWITCH);
+	// variable stating the current tilt state of the acquisition - only has 'up' and 'down' values
+	private String _tiltState = "up";
```
Comment (vladaviedov) [inline src/main/java/frc/robot/subsystems/Acquisition.java:17]: This should be a boolean

File: src/main/java/frc/robot/subsystems/Acquisition.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	public boolean getTiltLimit() {
+		return _tiltLimitSwitch.get();
+	}
+
+	/**
+	 * returns if the acquisition is spinning (this is used for dashboard)
+	 * @return acquisition spinning (true = spinning, false = not spinning)
+	 */
+	public boolean getSpinState() {
+		return _spinMotor.get() > 0;
```
Comment (vladaviedov) [inline src/main/java/frc/robot/subsystems/Acquisition.java:76]: This does not handle negative values

File: src/main/java/frc/robot/subsystems/Acquisition.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+			_tiltMotor.set(0);
+		}
+	}
+
+	/**
+	 * returns if the limit switch has been activated (and the acquisition can't tilt farther down)
+	 * @return the value of the limit switch (true is pressed, false is not pressed)
+	 */
+	public boolean getTiltLimit() {
+		return _tiltLimitSwitch.get();
```
Comment (vladaviedov) [inline src/main/java/frc/robot/subsystems/Acquisition.java:68]: Limit switches returns true when not pressed, this behavior will be inconsistent with the comment

File: src/main/java/frc/robot/subsystems/Acquisition.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		_tiltState = newState;
+	}
+
+	/**
+	 * spins spin motor at 'speed' speed and ball channel motor at '-5/3 * speed' speed
+	 * @param speed base value for speed calculations, always >0
+	 */
+	public void spinAcquisition(double speed) {
+		_spinMotor.set(speed);
+		_ballChannelMotor.set(-speed * 5 / 3);
```
Comment (vladaviedov) [inline src/main/java/frc/robot/subsystems/Acquisition.java:48]: Magic number over here, should have a local constant for this

File: src/main/java/frc/robot/subsystems/Acquisition.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	/**
+	 * returns the state of the acquisition tilt
+	 * @return true if up, false if down
+	 */
+	public boolean getTiltState() {
+		return "up".equals(_tiltState);
+	}
+
```
Comment (vladaviedov) [inline src/main/java/frc/robot/subsystems/Acquisition.java:32]: Where will these two functions (get/setTitlState) be used in the program?
Why are we changing the tiltState from outside?

File: src/main/java/frc/robot/subsystems/Acquisition.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	public void spinAcquisition(double speed) {
+		_spinMotor.set(speed);
+		_ballChannelMotor.set(-speed * 5 / 3);
+	}
+
+	/**
+	 * spins tilt motor at 'speed' speed if it's raising it or lowering it and hasn't hit the limit switch
+	 * @param speed speed of motor
+	 */
+	public void tiltAcquisition(double speed) {
```
Comment (vladaviedov) [inline src/main/java/frc/robot/subsystems/Acquisition.java:55]: Name is a bit too verbose. You're calling it on an Acquisition object, so it's obvious what it's going to tilt

File: src/main/java/frc/robot/commands/manual/ManualAcquisition.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+			_acquisition.tiltAcquisition(TILT_SPEED);
+		} else if (_controls.getAcqTilt() < 0) {
+			_acquisition.tiltAcquisition(-TILT_SPEED);
+		} else {
+			_acquisition.tiltAcquisition(0);
+		}
+	}
+
+	@Override
+	public boolean isFinished() {
```
Comment (vladaviedov) [inline src/main/java/frc/robot/commands/manual/ManualAcquisition.java:42]: This override is not needed, isFinished returns false by default

**Review by vladaviedov — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Acquisition.java (hunk 7 above)
Comment (fielia) [reply src/main/java/frc/robot/subsystems/Acquisition.java:32]: auto commands (TiltAcquisition.java in the old code)

File: src/main/java/frc/robot/RobotMap.java (hunk 1 above)
Comment (vladaviedov) [reply src/main/java/frc/robot/RobotMap.java:20]: The rest are CAN ports, only the limit switch is a PWM

File: src/main/java/frc/robot/subsystems/Acquisition.java (hunk 7 above)
Comment (vladaviedov) [reply src/main/java/frc/robot/subsystems/Acquisition.java:34]: Then this should be based on the limit switch value instead

**Review by vladaviedov — CHANGES_REQUESTED** [review]
Still some issues left
Don't forget to sync the branch to dev and update the package name on your files

**Review by vladaviedov — APPROVED** [review]

**Pull request merged**